src/supervisor.rs

```diff
@@ -0,0 +1,614 @@
+    let mut children_specs = Vec::new();
+    let mut termination_errors: HashMap<String, Arc<worker::TerminationError>> = HashMap::new();
+    for runtime_child in runtime_children {
+        let child_runtime_name = runtime_child.runtime_name.to_owned();
+        let result = runtime_child.terminate().await;
+        match result {
+            Err((failed_child_spec, termination_err0)) => {
+                let termination_err = Arc::new(termination_err0);
+                ev_notifier
+                    .worker_termination_failed(child_runtime_name.clone(), termination_err.clone())
+                    .await;
+                termination_errors.insert(child_runtime_name, termination_err);
+                children_specs.push(failed_child_spec);
+            }
+            Ok(failed_child_spec) => {
+                ev_notifier
+                    .worker_terminated(child_runtime_name.clone())
+                    .await;
+                children_specs.push(failed_child_spec);
+            }
+        }
+    }
+    if termination_errors.is_empty() {
+        Ok(children_specs)
+    } else {
+        Err((children_specs, TerminationError(termination_errors)))
+    }
+}
+
+impl Spec {
+    pub fn new(
+        name: impl Into<String>,
+        children: Vec<worker::Spec>,
+        ev_notifier: EventNotifier,
+    ) -> Spec {
+        Spec {
+            name: name.into(),
+            ev_notifier,
+            children,
+
+            order: Order::LeftToRight,
+            strategy: Strategy::OneForOne,
+        }
+    }
+
+    async fn start_supervisor<'a>(
+        self,
+        parent_ctx: &context::Context,
+        parent_name: &'a str,
+    ) -> Result<Supervisor, (Spec, Arc<StartError>)> {
+        let (ctx, _terminate_supervisor) = Context::with_cancel(parent_ctx);
+
+        let name = self.name.clone();
+        let mut ev_notifier = self.ev_notifier.clone();
+        let order = self.order.clone();
+        let strategy = self.strategy.clone();
+
+        let sup_runtime_name = format!("{}/{}", parent_name, name);
+        let result = start_children(
+            &ctx,
+            ev_notifier.clone(),
+            &sup_runtime_name,
+            &order,
+            self.children,
+        )
+        .await;
+
+        match result {
+            Err((children_specs, err0)) => {
+                let err = Arc::new(err0);
+                ev_notifier
+                    .supervisor_start_failed(&sup_runtime_name, err.clone())
+                    .await;
+                let spec = Spec {
+                    name,
+                    ev_notifier,
+                    order,
+                    strategy,
+                    children: children_specs,
+                };
+                return Err((spec, err));
+            }
+            Ok(children) => {
+                ev_notifier.supervisor_started(&sup_runtime_name).await;
+                Ok(Supervisor {
+                    runtime_name: sup_runtime_name.to_owned(),
```

Comment:
I'm pretty sure `sup_runtime_name` is already a `String`.

```suggestion
                    runtime_name: sup_runtime_name,
```